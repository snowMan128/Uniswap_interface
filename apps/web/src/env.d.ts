// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../../../index.d.ts" />
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../../../packages/wallet/src/react-native-dotenv.d.ts" />
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../../../packages/ui/src/env.d.ts" />

import { tamaguiConfig } from 'ui/src'

type Conf = typeof tamaguiConfig

declare module 'tamagui' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface TamaguiCustomConfig extends Conf {}
}
