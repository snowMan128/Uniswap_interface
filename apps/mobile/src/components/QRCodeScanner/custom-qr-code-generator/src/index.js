// Component logic from: https://github.com/awesomejerry/react-native-qrcode-svg
// Custom matric renderer from: https://github.com/awesomejerry/react-native-qrcode-svg/pull/139/files

import React, { useMemo } from 'react'
import Svg, { Defs, G, LinearGradient, Path, Rect, Stop } from 'react-native-svg'
import genMatrix from 'src/components/QRCodeScanner/custom-qr-code-generator/src/genMatrix.js'
import transformMatrixIntoPath from 'src/components/QRCodeScanner/custom-qr-code-generator/src/transformMatrixIntoCirclePath.js'
import { useMedia } from 'ui/src'

const QREyes = ({ x = -1, y = -1, fillColor, size }) => (
  <G transform={`scale(${size / 120})`} x={x} y={y}>
    <Path
      clip-rule="evenodd"
      d="M0 12C0 5.37258 5.37258 0 12 0H28C34.6274 0 40 5.37258 40 12V28C40 34.6274 34.6274 40 28 40H12C5.37258 40 0 34.6274 0 28V12ZM28 6.27451H12C8.8379 6.27451 6.27451 8.8379 6.27451 12V28C6.27451 31.1621 8.8379 33.7255 12 33.7255H28C31.1621 33.7255 33.7255 31.1621 33.7255 28V12C33.7255 8.8379 31.1621 6.27451 28 6.27451Z"
      fill={fillColor}
      fill-rule="evenodd"
    />
    <Path
      d="M11 17C11 13.6863 13.6863 11 17 11H23C26.3137 11 29 13.6863 29 17V23C29 26.3137 26.3137 29 23 29H17C13.6863 29 11 26.3137 11 23V17Z"
      fill={fillColor}
    />
  </G>
)

const QREyeBG = ({ x = -1, y = -1, size, backgroundColor }) => (
  <G transform={`scale(${size / 120})`} x={x} y={y}>
    <Path d="M0 0H40V40H0V0Z" fill={backgroundColor} />
  </G>
)

const QREyeWrapper = ({ x = 0, y = 0, backgroundColor, overlayColor, fillColor, size }) => (
  <>
    <QREyeBG backgroundColor={backgroundColor} size={size} x={x} y={y} />
    <QREyes fillColor={fillColor} size={size} x={x} y={y} />
    <QREyes fillColor={overlayColor} size={size} x={x} y={y} />
  </>
)

const QRCode = ({
  value = 'Wallet QR code',
  size = 190,
  color,
  backgroundColor,
  overlayColor = '#FFFFFF',
  borderRadius = 24,
  quietZone = 8,
  enableLinearGradient = false,
  gradientDirection = ['0%', '0%', '100%', '100%'],
  linearGradient = ['rgb(255,255,255)', 'rgb(0,255,255)'],
  ecl = 'H',
  getRef,
  onError,
}) => {
  const result = useMemo(() => {
    try {
      return transformMatrixIntoPath(genMatrix(value, ecl), size)
    } catch (error) {
      if (onError && typeof onError === 'function') {
        onError(error)
      } else {
        // Pass the error when no handler presented
        throw error
      }
    }
  }, [value, size, ecl, onError])

  const media = useMedia()
  if (!result) {
    return null
  }

  const { path } = result

  const eyeSize = media.short ? 126 : 138

  return (
    <Svg
      ref={getRef}
      height={size}
      viewBox={[-quietZone, -quietZone, size + quietZone * 2, size + quietZone * 2].join(' ')}
      width={size}>
      <Defs>
        <LinearGradient
          gradientTransform="rotate(45)"
          id="grad"
          x1={gradientDirection[0]}
          x2={gradientDirection[2]}
          y1={gradientDirection[0]}
          y2={gradientDirection[2]}>
          <Stop offset="0" stopColor={color} stopOpacity="1" />
          <Stop offset="1" stopColor={linearGradient[1]} stopOpacity="1" />
          <Stop offset="1" stopColor={linearGradient[2]} stopOpacity="1" />
        </LinearGradient>
      </Defs>
      <G>
        <Rect
          fill={backgroundColor}
          height={size + quietZone * 2}
          rx={borderRadius}
          width={size + quietZone * 2}
          x={-quietZone}
          y={-quietZone}
        />
      </G>
      <G>
        <Path d={path} fill={enableLinearGradient ? 'url(#grad)' : color} />
        <Path d={path} fill={enableLinearGradient ? overlayColor + '66' : overlayColor + '2D'} />
        <QREyeWrapper
          backgroundColor={backgroundColor}
          fillColor={color}
          overlayColor={overlayColor + '2D'}
          size={eyeSize}
        />
        <QREyeWrapper
          backgroundColor={backgroundColor}
          fillColor={color}
          overlayColor={overlayColor + '2D'}
          size={eyeSize}
          y={size - eyeSize / 3}
        />
        <QREyeWrapper
          backgroundColor={backgroundColor}
          fillColor={color}
          overlayColor={overlayColor + '2D'}
          size={eyeSize}
          x={size - eyeSize / 3}
        />
      </G>
    </Svg>
  )
}

export default QRCode
