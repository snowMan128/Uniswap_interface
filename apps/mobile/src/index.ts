// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../../../index.d.ts" />

// TODO - remove this declaration after updating react-native-reanimted
// to the never version (3.3.0 has typescript type issues and is missing
// this declaration in the react-native-reanimated.d.ts file)
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toHaveAnimatedStyle(
        style: Record<string, unknown>[] | Record<string, unknown>,
        config?: {
          shouldMatchAllProps?: boolean
        }
      ): R
    }
  }
}

export {}
