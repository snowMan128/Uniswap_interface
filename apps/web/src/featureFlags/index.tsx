import useParsedQueryString from 'hooks/useParsedQueryString'
import { atomWithStorage, useAtomValue, useUpdateAtom } from 'jotai/utils'
import { ReactNode, createContext, useCallback, useContext, useEffect, useMemo } from 'react'
import { useGate } from 'statsig-react'
import { isDevelopmentEnv, isStagingEnv } from 'utils/env'

/**
 * The value here must match the value in the statsig dashboard, if you plan to use statsig.
 */
export enum FeatureFlag {
  realtime = 'realtime',
  traceJsonRpc = 'traceJsonRpc',
  fallbackProvider = 'fallback_provider',
  uniswapXSyntheticQuote = 'uniswapx_synthetic_quote',
  multichainUX = 'multichain_ux',
  currencyConversion = 'currency_conversion',
  quickRouteMainnet = 'enable_quick_route_mainnet',
  feesEnabled = 'fees_enabled',
  limitsEnabled = 'limits_enabled',
  eip6963Enabled = 'eip6963_enabled',
  gatewayDNSUpdate = 'gateway_dns_update',
  sendEnabled = 'swap_send',
  gatewayDNSUpdateAll = 'gateway_dns_update_all',
  uniconV2 = 'unicon_V2',
  limitsFees = 'limits_fees',
  exitAnimation = 'exit_animation',
  v2Everywhere = 'v2_everywhere',
  gqlTokenLists = 'gql_token_lists',
  // TODO(WEB-3625): Remove these once we have a generalized system for outage banners.
  outageBannerOptimism = 'outage_banner_feb_2024_optimism',
  outageBannerArbitrum = 'outage_banner_feb_2024_arbitrum',
  outageBannerPolygon = 'outage_banner_feb_2024_polygon',
}

interface FeatureFlagsContextType {
  isLoaded: boolean
  flags: Record<string, string>
  configs: Record<string, any>
}

const FeatureFlagContext = createContext<FeatureFlagsContextType>({ isLoaded: false, flags: {}, configs: {} })

export function useFeatureFlagsContext(): FeatureFlagsContextType {
  const context = useContext(FeatureFlagContext)
  if (!context) {
    throw Error('Feature flag hooks can only be used by children of FeatureFlagProvider.')
  } else {
    return context
  }
}

/* update and save feature flag & dynamic config settings */
export const featureFlagSettings = atomWithStorage<Record<string, string>>('featureFlags', {})
export const dynamicConfigSettings = atomWithStorage<Record<string, any>>('dynamicConfigs', {})

export function useUpdateFlag() {
  const setFeatureFlags = useUpdateAtom(featureFlagSettings)

  return useCallback(
    (featureFlag: string, option: string) => {
      setFeatureFlags((featureFlags) => ({
        ...featureFlags,
        [featureFlag]: option,
      }))
    },
    [setFeatureFlags]
  )
}

export function useUpdateConfig() {
  const setConfigs = useUpdateAtom(dynamicConfigSettings)

  return useCallback(
    (configName: string, option: any) => {
      setConfigs((configs) => ({
        ...configs,
        [configName]: option,
      }))
    },
    [setConfigs]
  )
}

export function FeatureFlagsProvider({ children }: { children: ReactNode }) {
  // TODO: `isLoaded` to `true` so `App.tsx` will render. Later, this will be dependent on
  // flags loading from Amplitude, with a timeout.
  const featureFlags = useAtomValue(featureFlagSettings)
  const dynamicConfigs = useAtomValue(dynamicConfigSettings)
  const value = {
    isLoaded: true,
    flags: featureFlags,
    configs: dynamicConfigs,
  }
  return <FeatureFlagContext.Provider value={value}>{children}</FeatureFlagContext.Provider>
}

export function useFeatureFlagsIsLoaded(): boolean {
  return useFeatureFlagsContext().isLoaded
}

export enum BaseVariant {
  Control = 'control',
  Enabled = 'enabled',
}

export function useBaseFlag(flag: string, defaultValue = BaseVariant.Control): BaseVariant {
  const { value: statsigValue } = useGate(flag) // non-existent gates return false
  const featureFlagsContext = useFeatureFlagsContext()
  if (statsigValue) {
    return BaseVariant.Enabled
  }
  switch (featureFlagsContext.flags[flag]) {
    case 'enabled':
      return BaseVariant.Enabled
    case 'control':
      return BaseVariant.Control
    default:
      return defaultValue
  }
}

export function useFeatureFlagURLOverrides() {
  const parsedQs = useParsedQueryString()
  const setFeatureFlags = useUpdateAtom(featureFlagSettings)

  const newFeatureFlagObj = useMemo(() => {
    const featureFlagOverrides =
      typeof parsedQs.featureFlagOverride === 'string' ? parsedQs.featureFlagOverride.split(',') : []
    return featureFlagOverrides.reduce(
      (prev, current) => ({
        ...prev,
        [current]: BaseVariant.Enabled,
      }),
      {}
    )
  }, [parsedQs])

  useEffect(() => {
    if (!isDevelopmentEnv() && !isStagingEnv()) return

    setFeatureFlags((prev) => ({
      ...prev,
      ...newFeatureFlagObj,
    }))
  }, [newFeatureFlagObj, setFeatureFlags])
}
