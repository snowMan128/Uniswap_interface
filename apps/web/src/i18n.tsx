import { useEffect } from 'react'

import { initialLocale, useActiveLocale } from 'hooks/useActiveLocale'
import { ReactNode } from 'react'
import { useUserLocaleManager } from 'state/user/hooks'

import i18n, { t } from 'i18next'
import { Trans as OGTrans, Translation, initReactI18next, useTranslation as useTranslationOG } from 'react-i18next'

import { SupportedLocale } from 'constants/locales'
import resourcesToBackend from 'i18next-resources-to-backend'

export { t } from 'i18next'

export const Trans = ((props) => {
  // forces re-render on language change because it doesn't by default
  useTranslation()
  return <OGTrans {...props}>{props.children}</OGTrans>
}) satisfies typeof OGTrans

export function useTranslation() {
  if (process.env.NODE_ENV === 'test') {
    return { i18n, t }
  }
  // eslint-disable-next-line react-hooks/rules-of-hooks
  return useTranslationOG()
}

export function Plural({ value, one, other }: { value: number; one: string; other: string }) {
  const children = value === 1 ? one : other
  if (process.env.NODE_ENV === 'test') {
    return <>{children}</>
  }
  // ensures it re-renders when language changes
  return <Translation>{() => children}</Translation>
}

i18n
  .use(initReactI18next)
  .use(
    resourcesToBackend((language: string, namespace: string) => {
      // not sure why but it tries to load es THEN es-ES, for any language, but we just want the second
      if (!language.includes('-')) {
        return
      }
      if (language === 'en-US') {
        if (process.env.NODE_ENV === 'test') {
          return import('./i18n/locales/source/en-US.json')
        }
      }
      return import(`./i18n/locales/${namespace}/${language}.json`)
    })
  )
  .on('failedLoading', (language, namespace, msg) => {
    console.error(`Error loading language ${language} ${namespace}: ${msg}`)
  })

i18n
  .init({
    returnEmptyString: false,
    defaultNS: 'translations',
    keySeparator: '~~',
    lng: 'en-US',
    fallbackLng: 'en-US',
    interpolation: {
      escapeValue: false, // react already safes from xss
    },
    react: {
      transSupportBasicHtmlNodes: true,
    },
  })
  .catch(() => undefined)

let changingTo = ''

async function dynamicActivate(locale: SupportedLocale) {
  if (i18n.language === locale || locale === changingTo) return
  // since its async we need to "lock" while its changing
  changingTo = locale
  await i18n.changeLanguage(locale)
  i18n.emit('')
  changingTo = ''
}

dynamicActivate(initialLocale)

export function LanguageProvider({ children }: { children: ReactNode }): JSX.Element {
  const activeLocale = useActiveLocale()
  const [userLocale, setUserLocale] = useUserLocaleManager()
  const locale = userLocale || activeLocale

  useEffect(() => {
    dynamicActivate(locale)
    document.documentElement.setAttribute('lang', locale)
    setUserLocale(locale) // stores the selected locale to persist across sessions
  }, [setUserLocale, locale])

  return <>{children}</>
}
