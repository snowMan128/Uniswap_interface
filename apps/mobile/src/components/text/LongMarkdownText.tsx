import React, { useCallback, useReducer, useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { LayoutChangeEvent } from 'react-native'
import Markdown, { MarkdownProps } from 'react-native-markdown-display'
import { Flex, SpaceTokens, Text, useSporeColors } from 'ui/src'
import { fonts } from 'ui/src/theme'
import { openUri } from 'wallet/src/utils/linking'

type LongMarkdownTextProps = {
  initialDisplayedLines?: number
  text: string
  gap?: SpaceTokens
  color?: string
  linkColor?: string
  codeBackgroundColor?: string
  readMoreOrLessColor?: string
  variant?: keyof typeof fonts
}

export function LongMarkdownText(props: LongMarkdownTextProps): JSX.Element {
  const colors = useSporeColors()
  const { t } = useTranslation()
  const {
    initialDisplayedLines = 3,
    text,
    gap = '$spacing8',
    color = colors.neutral1.val,
    linkColor = colors.neutral2.val,
    readMoreOrLessColor = colors.neutral2.val,
    codeBackgroundColor = colors.surface3.val,
    variant = 'body2',
  } = props

  const [expanded, toggleExpanded] = useReducer((isExpanded) => !isExpanded, true)
  const [textLengthExceedsLimit, setTextLengthExceedsLimit] = useState(false)
  const [textLineHeight, setTextLineHeight] = useState<number>(fonts[variant].lineHeight)
  const initialContentHeightRef = useRef<number>()
  const maxVisibleHeight = textLineHeight * initialDisplayedLines

  const onMarkdownLayout = useCallback(
    (event: LayoutChangeEvent) => {
      if (initialContentHeightRef.current !== undefined) {
        return
      }
      const textContentHeight = event.nativeEvent.layout.height
      const currentLines = Math.floor(textContentHeight / textLineHeight)
      setTextLengthExceedsLimit(currentLines > initialDisplayedLines)
      toggleExpanded()
      initialContentHeightRef.current = textContentHeight
    },
    [initialDisplayedLines, textLineHeight]
  )

  const codeStyle = { backgroundColor: codeBackgroundColor, borderColor: 'transparent' }

  const markdownStyle: MarkdownProps['style'] = {
    body: {
      color,
      fontFamily: fonts[variant].family,
      overflow: 'hidden',
      height: !textLengthExceedsLimit || expanded ? 'auto' : maxVisibleHeight,
    },
    code_inline: codeStyle,
    fence: codeStyle,
    code_block: codeStyle,
    link: { color: linkColor },
    paragraph: {
      marginBottom: 0,
      marginTop: 0,
      fontSize: fonts.body2.fontSize,
      lineHeight: textLineHeight,
    },
  }

  return (
    <Flex gap={gap}>
      <Flex testID="markdown-wrapper" onLayout={onMarkdownLayout}>
        {/* Render fake one-line markdown to properly measure the height of a single text line */}
        <Flex
          opacity={0}
          pointerEvents="none"
          position="absolute"
          onLayout={({
            nativeEvent: {
              layout: { height },
            },
          }): void => {
            setTextLineHeight(height)
          }}>
          <Markdown
            style={{
              ...markdownStyle,
              body: { ...markdownStyle.body, height: 'auto' },
              paragraph: { ...markdownStyle.paragraph, lineHeight: fonts[variant].lineHeight },
            }}
            {...{ children: '.' }}
          />
        </Flex>
        <Markdown
          style={markdownStyle}
          onLinkPress={(url): false => {
            // add our own custom link handler since it has security checks that only open http/https links
            openUri(url).catch(() => undefined)
            return false
          }}
          // HACK: children prop no in TS definition
          {...{ children: text }}
        />
      </Flex>

      {/* Text is removed vs hidden using opacity to ensure spacing after the element is consistent in all cases.
      This will cause mild thrash as data loads into a page but will ensure consistent spacing */}
      {textLengthExceedsLimit ? (
        <Text
          my="$none"
          py="$none"
          style={{ color: readMoreOrLessColor }}
          testID="read-more-button"
          variant="buttonLabel3"
          onPress={toggleExpanded}>
          {expanded ? t('Read less') : t('Read more')}
        </Text>
      ) : null}
    </Flex>
  )
}
