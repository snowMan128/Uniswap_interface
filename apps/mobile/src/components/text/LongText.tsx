import React, { ComponentProps, useCallback, useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { NativeSyntheticEvent, TextLayoutEventData } from 'react-native'
import { Flex, SpaceTokens, Text, useSporeColors } from 'ui/src'
import { fonts } from 'ui/src/theme'

type LongTextProps = {
  initialDisplayedLines?: number
  text: string
  gap?: SpaceTokens
  color?: string
  linkColor?: string
  codeBackgroundColor?: string
  readMoreOrLessColor?: string
  variant?: keyof typeof fonts
} & Omit<
  ComponentProps<typeof Text>,
  'children' | 'numberOfLines' | 'onTextLayout' | 'color' | 'variant'
>

export function LongText(props: LongTextProps): JSX.Element {
  const colors = useSporeColors()
  const { t } = useTranslation()
  const {
    initialDisplayedLines = 3,
    text,
    gap = '$spacing8',
    color = colors.neutral1.val,
    readMoreOrLessColor = colors.neutral2.val,
    variant = 'body2',
    ...rest
  } = props

  const [expanded, setExpanded] = useState(true)
  const [textLengthExceedsLimit, setTextLengthExceedsLimit] = useState(false)
  const isInitializedRef = useRef(false)

  const onTextLayout = useCallback(
    (e: NativeSyntheticEvent<TextLayoutEventData>) => {
      // Only needs to measure full number of lines once
      if (isInitializedRef.current) {
        return
      }
      setTextLengthExceedsLimit(e.nativeEvent.lines.length > initialDisplayedLines)
      setExpanded(false)
      isInitializedRef.current = true
    },
    [initialDisplayedLines]
  )

  return (
    <Flex gap={gap}>
      <Text
        numberOfLines={expanded ? undefined : initialDisplayedLines}
        style={{ color }}
        variant={variant}
        onTextLayout={onTextLayout}
        {...rest}>
        {text}
      </Text>

      {/* Text is removed vs hidden using opacity to ensure spacing after the element is consistent in all cases.
      This will cause mild thrash as data loads into a page but will ensure consistent spacing */}
      {textLengthExceedsLimit ? (
        <Text
          my="$none"
          py="$none"
          style={{ color: readMoreOrLessColor }}
          testID="read-more-button"
          variant="buttonLabel3"
          onPress={(): void => setExpanded(!expanded)}>
          {expanded ? t('Read less') : t('Read more')}
        </Text>
      ) : null}
    </Flex>
  )
}
