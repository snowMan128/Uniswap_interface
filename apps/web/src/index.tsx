/* eslint-disable prettier/prettier */
// Ordering is intentional and must be preserved: styling, polyfilling, tracing, and then functionality.
import '@reach/dialog/styles.css'
import 'inter-ui'
import 'polyfills'
import 'tracing'
import 'connection/eagerlyConnect'
/* eslint-enable prettier/prettier */

import { FeatureFlagsProvider } from 'featureFlags'
import { Provider as ApolloProvider } from 'graphql/data/apollo/Provider'
import { BlockNumberProvider } from 'lib/hooks/useBlockNumber'
import { MulticallUpdater } from 'lib/state/multicall'
import { StrictMode } from 'react'
import { createRoot } from 'react-dom/client'
import { Helmet, HelmetProvider } from 'react-helmet-async/lib/index'
import { QueryClient, QueryClientProvider } from 'react-query'
import { Provider } from 'react-redux'
import { BrowserRouter, HashRouter, useLocation } from 'react-router-dom'
import { SystemThemeUpdater, ThemeColorMetaUpdater } from 'theme/components/ThemeToggle'
import { TamaguiProvider } from 'theme/tamaguiProvider'
import { UnitagUpdaterContextProvider } from 'uniswap/src/features/unitags/context'
import { isBrowserRouterEnabled } from 'utils/env'
import { getCanonicalUrl } from 'utils/urlRoutes'
import Web3Provider from './components/Web3Provider'
import { LanguageProvider } from './i18n'
import App from './pages/App'
import * as serviceWorkerRegistration from './serviceWorkerRegistration'
import store from './state'
import ApplicationUpdater from './state/application/updater'
import ListsUpdater from './state/lists/updater'
import LogsUpdater from './state/logs/updater'
import OrderUpdater from './state/signatures/updater'
import TransactionUpdater from './state/transactions/updater'
import { ThemeProvider, ThemedGlobalStyle } from './theme'
import RadialGradientByChainUpdater from './theme/components/RadialGradientByChainUpdater'

if (window.ethereum) {
  window.ethereum.autoRefreshOnNetworkChange = false
}

function Updaters() {
  const location = useLocation()

  return (
    <>
      <Helmet>
        <link rel="canonical" href={getCanonicalUrl(location.pathname)} />
      </Helmet>
      <RadialGradientByChainUpdater />
      <ListsUpdater />
      <SystemThemeUpdater />
      <ThemeColorMetaUpdater />
      <ApplicationUpdater />
      <TransactionUpdater />
      <OrderUpdater />
      <MulticallUpdater />
      <LogsUpdater />
    </>
  )
}

const queryClient = new QueryClient()

const container = document.getElementById('root') as HTMLElement

const Router = isBrowserRouterEnabled() ? BrowserRouter : HashRouter

createRoot(container).render(
  <StrictMode>
    <HelmetProvider>
      <Provider store={store}>
        <FeatureFlagsProvider>
          <QueryClientProvider client={queryClient}>
            <Router>
              <LanguageProvider>
                <Web3Provider>
                  <ApolloProvider>
                    <BlockNumberProvider>
                      <UnitagUpdaterContextProvider>
                        <Updaters />
                        <ThemeProvider>
                          <TamaguiProvider>
                            <ThemedGlobalStyle />
                            <App />
                          </TamaguiProvider>
                        </ThemeProvider>
                      </UnitagUpdaterContextProvider>
                    </BlockNumberProvider>
                  </ApolloProvider>
                </Web3Provider>
              </LanguageProvider>
            </Router>
          </QueryClientProvider>
        </FeatureFlagsProvider>
      </Provider>
    </HelmetProvider>
  </StrictMode>
)

if (process.env.REACT_APP_SERVICE_WORKER !== 'false') {
  serviceWorkerRegistration.register()
}
