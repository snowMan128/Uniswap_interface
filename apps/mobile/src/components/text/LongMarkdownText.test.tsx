import React from 'react'
import { MarkdownProps } from 'react-native-markdown-display'
import { ReactTestInstance } from 'react-test-renderer'
import { LongMarkdownText } from 'src/components/text/LongMarkdownText'
import { fireEvent, render, within } from 'src/test/test-utils'
import { fonts } from 'ui/src/theme'

const TEXT_VARIANT = 'body2'
const LINE_HEIGHT = fonts[TEXT_VARIANT].lineHeight

const SHORT_TEXT = 'Short text'
const LONG_TEXT = 'Some very long text'

jest.mock('react-native-markdown-display', () => {
  const Markdown = jest.requireActual('react-native-markdown-display').default

  return {
    __esModule: true, // this property makes Markdown renderering work in the es module
    default: jest.fn().mockImplementation((props: MarkdownProps) => <Markdown {...props} />),
  }
})

const fireLayoutEvent = (instance: ReactTestInstance, lines: number): void => {
  const height = lines * LINE_HEIGHT

  fireEvent(instance, 'layout', {
    nativeEvent: {
      layout: {
        height,
        width: 100,
      },
    },
  })
}

const renderMarkdown = (text: string): ReturnType<typeof render> =>
  render(<LongMarkdownText initialDisplayedLines={3} text={text} variant={TEXT_VARIANT} />)

const measureMarkdown = (tree: ReturnType<typeof render>, numberOfLines: number): void => {
  const markdownWrapperInstance = tree.getByTestId('markdown-wrapper')
  fireLayoutEvent(markdownWrapperInstance, numberOfLines)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getMarkdownPropsWithHeight = (height: number | 'auto'): any =>
  expect.objectContaining({
    style: expect.objectContaining({
      body: expect.objectContaining({
        height, // height auto means the text doesn't exceed the limit
      }),
    }),
  })

describe(LongMarkdownText, () => {
  const MockedMarkdown = require('react-native-markdown-display').default as jest.Mock

  it('renders without error', () => {
    const tree = renderMarkdown(LONG_TEXT)

    expect(tree).toMatchSnapshot()
  })

  describe('short text not exceeding the limit', () => {
    it('shows the entire text', () => {
      const tree = renderMarkdown(SHORT_TEXT)
      measureMarkdown(tree, 1) // Assume Short text is one line

      // props are at index 0, ref is at index 1
      expect(MockedMarkdown.mock.lastCall[0]).toEqual(
        getMarkdownPropsWithHeight('auto') // height auto means the text doesn't exceed the limit
      )
    })

    it('does not display the "read more" button', () => {
      const tree = renderMarkdown(SHORT_TEXT)
      measureMarkdown(tree, 1) // Assume Short text is one line

      const readMoreButton = tree.queryByTestId('read-more-button')

      expect(readMoreButton).toBeNull()
    })
  })

  describe('long text exceeding the limit', () => {
    describe('when the text is not expanded', () => {
      it('limits the number of visible lines', () => {
        const tree = renderMarkdown(LONG_TEXT)

        measureMarkdown(tree, 5) // Assume Some very long text is five lines

        expect(MockedMarkdown.mock.lastCall[0]).toEqual(
          getMarkdownPropsWithHeight(LINE_HEIGHT * 3) // Height is limited to 3 lines
        )
      })

      it('displays the "read more" button', () => {
        const tree = renderMarkdown(LONG_TEXT)

        measureMarkdown(tree, 5) // Assume Some very long text is five lines

        const readMoreButton = tree.queryByTestId('read-more-button')

        expect(readMoreButton).toBeTruthy()
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        expect(within(readMoreButton!).getByText('Read more')).toBeTruthy()
      })
    })

    describe('when the text is expanded', () => {
      it('shows the entire text', () => {
        const tree = renderMarkdown(LONG_TEXT)

        measureMarkdown(tree, 5) // Assume Some very long text is five lines

        const readMoreButton = tree.getByTestId('read-more-button')
        fireEvent.press(readMoreButton)

        expect(MockedMarkdown.mock.lastCall[0]).toEqual(
          getMarkdownPropsWithHeight('auto') // height auto means the text doesn't exceed the limit
        )
      })

      it('displays the "read less" button', () => {
        const tree = renderMarkdown(LONG_TEXT)

        measureMarkdown(tree, 5) // Assume Some very long text is five lines

        const readMoreButton = tree.getByTestId('read-more-button')
        fireEvent.press(readMoreButton)

        expect(readMoreButton).toBeTruthy()
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        expect(within(readMoreButton!).getByText('Read less')).toBeTruthy()
      })
    })

    it('toggles the text when the "read more/less" button is pressed', () => {
      const tree = renderMarkdown(LONG_TEXT)

      measureMarkdown(tree, 5) // Assume Some very long text is five lines

      const readMoreButton = tree.getByTestId('read-more-button')
      fireEvent.press(readMoreButton) // expand

      expect(MockedMarkdown.mock.lastCall[0]).toEqual(
        getMarkdownPropsWithHeight('auto') // height auto means the text doesn't exceed the limit
      )

      fireEvent.press(readMoreButton) // collapse

      expect(MockedMarkdown.mock.lastCall[0]).toEqual(
        getMarkdownPropsWithHeight(LINE_HEIGHT * 3) // Height is limited to 3 lines
      )
    })
  })
})
