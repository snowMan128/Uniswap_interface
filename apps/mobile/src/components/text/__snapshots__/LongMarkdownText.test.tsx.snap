// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LongMarkdownText renders without error 1`] = `
<View
  style={
    {
      "alignItems": "stretch",
      "flexDirection": "column",
      "gap": 8,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "stretch",
        "flexDirection": "column",
      }
    }
    testID="markdown-wrapper"
  >
    <View
      onLayout={[Function]}
      pointerEvents="none"
      style={
        {
          "alignItems": "stretch",
          "flexDirection": "column",
          "opacity": 0,
          "position": "absolute",
        }
      }
    >
      <View
        style={
          {
            "height": "auto",
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "flex-start",
              "marginBottom": 0,
              "marginTop": 0,
              "width": "100%",
            }
          }
        >
          <Text
            style={{}}
          >
            <Text
              style={
                [
                  {
                    "color": "#222222",
                    "fontFamily": "Basel-Book",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  {},
                ]
              }
            >
              .
            </Text>
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "height": "auto",
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-start",
            "marginBottom": 0,
            "marginTop": 0,
            "width": "100%",
          }
        }
      >
        <Text
          style={{}}
        >
          <Text
            style={
              [
                {
                  "color": "#222222",
                  "fontFamily": "Basel-Book",
                  "fontSize": 17,
                  "lineHeight": 24,
                },
                {},
              ]
            }
          >
            Some very long text
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
